### ISSUE TITLE
TSQL: CREATE PROCEDURE does not support OUTPUT or GOTO

### ISSUE THREAD

**Issue by edmoren** [sqlfluff/sqlfluff#2948]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

While linting a stored procedure definition, the linter encountered a parser error. The error appears to be caused by the create statement. 

```
SET QUOTED_IDENTIFIER ON
GO
SET ANSI_NULLS ON
GO

CREATE PROCEDURE [dbo].[usp_process_tran_log]
	  @out_vchCode AS uddt_output_code OUTPUT
	, @out_vchMsg AS uddt_output_msg OUTPUT
	, @in_debug INT = 1
AS
--*******************************************************************************************
...
``` 

We encounter a parsing violation during linting. 
```
==== parsing violations ====
L:   4 | P:   3 |  PRS | Line 4, Position 3: Found unparsable section: '\n\nCREATE
                       | PROCEDURE [dbo].[usp_process_tr...'
```

### Expected Behaviour

Since `CREATE PROCEDURE` ... `AS` is valid `tsql`, I would expect to not encounter a parser error here.

### Observed Behaviour

As noted above, we get a parser violation. Checking the output of parse, yields this interesting output: 
```
...
[L:  4, P:  1]      |        keyword:                                              'GO'
[L:  4, P:  3]      |    unparsable:                                               !! Expected: 'Nothing...'
        Comments:
[L: 11, P:  1]      |            comment:                                          '--*******************************************************************************************'
        Code:
[L:  4, P:  3]      |            newline:                                          '\n'
[L:  5, P:  1]      |            newline:                                          '\n'
[L:  6, P:  1]      |            raw:                                              'CREATE'
...
```
Note the jump to Line 11 in the parser output followed by the return to line 4 before it continues. 

Removing the statements up to the `CREATE` results in no parser error.

## Basic Test 
Download `sample.txt` and rename to `sample.sql` 
Download `sqlfluff.txt` and rename to `.sqlfluff` or copy `.sqlfluff` config pasted below into a file called `.sqlfluff`. 
To view the raw parse output:
```
sqlfluff.exe parse ./sample.sql --dialect tsql --config ./.sqlfluff
``` 
 This should be the same as the `parse.txt` I've provided. 

Additionally, to view the parse exception in the lint output: 
```
sqlfluff.exe lint ./sample.sql --dialect tsql --config ./.sqlfluff
``` 

## Additional Test 
The removal of lines 6-10 in `sample.txt` will result in no parser error. Similarly, removal of lines 1-10 of `sample.txt` results in no parser error. This issue seems to be caused by the presence of the `tsql` definition keyword `CREATE PROCEDURE ... AS`

## Files
Example of `sqlfluff parse` output 
[parse.txt](https://github.com/sqlfluff/sqlfluff/files/8376745/parse.txt)
`sample.txt` will need to be renamed to `sample.sql` for the linter to work. Github rejects this file upload without renaming. 
[sample.txt](https://github.com/sqlfluff/sqlfluff/files/8376748/sample.txt)
`sqlfluff.txt` is the `.txt` version of my config file. Rename this to `.sqlfluff`. Github rejects this file upload without renaming. 
[sqlfluff.txt](https://github.com/sqlfluff/sqlfluff/files/8376750/sqlfluff.txt)

### Dialect

`tsql`

### Version

```
$ sqlfluff --version
sqlfluff, version 0.11.2
```

### Configuration

[sqlfluff]
dialect = tsql

[sqlfluff:indentation]
indented_joins = false
indented_using_on = true

[sqlfluff:rules]
indent_unit = tab
tab_space_size = 4
operator_new_lines = after
capitalisation_policy = upper
aliasing = explicit
allow_scalar = True
extended_capitalisation_policy = consistent
ignore_comment_clauses = True
ignore_comment_lines = True
max_line_length = 120
comma_style = leading
single_table_references = consistent
select_clause_trailing_comma = forbid
prefer_count_1 = True
multiline_newline = True
require_final_semicolon = True
allow_space_in_identifier = True
quoted_identifiers_policy = all
unquoted_identifiers_policy = all
exclude_rules = L025, L031, L061

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by tunetheweb** [comment]
SQLFluff current doesn't recognise this as a valid procedure parameter:

```
@out_vchCode AS uddt_output_code OUTPUT
```

There's two issues:
1. It doesn't understand "AS uddt_output_code". Is this a datatype? An alias? [It doesn't look like the T-SQL docs document this `AS` syntax](https://docs.microsoft.com/en-us/sql/t-sql/statements/create-procedure-transact-sql?view=sql-server-ver15).
2. It doesn't support the `OUTPUT` option. That bit is easily enough fixed.

**Comment by edmoren** [comment]
@tunetheweb 

1. It is valid syntax for `sql server` since you are allowed to define your own data-types. 

It is standard syntax for a create procedure. (Added emphasis **). 
```
-- Transact-SQL Syntax for Stored Procedures in SQL Server and Azure SQL Database

CREATE [ OR ALTER ] { PROC | PROCEDURE }
    [schema_name.] procedure_name [ ; number ]
    [ { @parameter_name [ type_schema_name. ] data_type }
        [ VARYING ] [ = default ] [ OUT | OUTPUT | [READONLY]
    ] [ ,...n ]
[ WITH <procedure_option> [ ,...n ] ]
[ FOR REPLICATION ]
**AS** { [ BEGIN ] sql_statement [;] [ ...n ] [ END ] }
[;]

<procedure_option> ::=
    [ ENCRYPTION ]
    [ RECOMPILE ]
    [ EXECUTE AS Clause ]
```

So a typical `CREATE PROCEDURE` with no frills might look like: 
```
CREATE PROCEDURE dbo.foo 
AS 
BEGIN 
SELECT 'bar'
END 
```

When I use a sample.sql as with standard-library datatypes and remove the `OUTPUT` hints as follows: 
```
SET QUOTED_IDENTIFIER ON
GO
SET ANSI_NULLS ON
GO

CREATE PROCEDURE [dbo].[usp_process_tran_log]
	  @out_vchCode AS NVARCHAR(10)
	, @out_vchMsg AS NVARCHAR(200)
	, @in_debug INT = 1
AS
--*******************************************************************************************
SET NOCOUNT ON;
BEGIN 
SELECT '8'
END 
```

I still get a parser error at the same line 
```
==== parsing violations ====
L:   4 | P:   3 |  PRS | Line 4, Position 3: Found unparsable section: '\n\nCREATE
                       | PROCEDURE [dbo].[usp_process_tr...
```

Raw parser: 
```
[L:  3, P: 18]      |    newline:                                                  '\n'
[L:  4, P:  1]      |    go_statement:
[L:  4, P:  1]      |        keyword:                                              'GO'
[L:  4, P:  3]      |    unparsable:                                               !! Expected: 'Nothing...'
        Comments:
[L: 11, P:  1]      |            comment:                                          '--*******************************************************************************************'        
        Code:
[L:  4, P:  3]      |            newline:                                          '\n'
[L:  5, P:  1]      |            newline:                                          '\n'
[L:  6, P:  1]      |            raw:                                              'CREATE'
[L:  6, P:  7]      |            whitespace:                                       ' '
[L:  6, P:  8]      |            raw:                                              'PROCEDURE'
[L:  6, P: 17]      |            whitespace:                                       ' '
[L:  6, P: 18]      |            raw:                                              '[dbo]'
[L:  6, P: 23]      |            raw:                                              '.'
[L:  6, P: 24]      |            raw:                                              '[usp_process_tran_log]'
[L:  6, P: 46]      |            newline:                                          '\n'
[L:  7, P:  1]      |            whitespace:                                       '\t  '
[L:  7, P:  4]      |            raw:                                              '@out_vchCode'
[L:  7, P: 16]      |            whitespace:                                       ' '
[L:  7, P: 17]      |            raw:                                              'AS'
[L:  7, P: 19]      |            whitespace:                                       ' '
[L:  7, P: 20]      |            raw:                                              'NVARCHAR'
[L:  7, P: 28]      |            raw:                                              '('
[L:  7, P: 29]      |            raw:                                              '10'
[L:  7, P: 31]      |            raw:                                              ')'
```

**Comment by tunetheweb** [comment]
> However, the `AS` is standard syntax for a create procedure. (Added emphasis **).

The `AS` you have emphasised is for creation of the procedure body, NOT the datatype in the procedure parameter.

SQLFluff accepts this:

```sql
CREATE PROCEDURE [dbo].[usp_process_tran_log]
	  @out_vchCode NVARCHAR(10)
	, @out_vchMsg NVARCHAR(200)
	, @in_debug INT = 1
AS
...
```

but not this:

```sql
CREATE PROCEDURE [dbo].[usp_process_tran_log]
	  @out_vchCode AS NVARCHAR(10)
	, @out_vchMsg AS NVARCHAR(200)
	, @in_debug INT = 1
AS
...
```


Similarly this parses fine in SQLFLuff:

```
CREATE PROCEDURE [dbo].[usp_process_tran_log]
	  @out_vchCode uddt_output_code
	, @out_vchMsg uddt_output_msg
	, @in_debug INT = 1
AS
```

**Comment by edmoren** [comment]
My mistake, I did not realize which `AS` you were referring to. 
The following is apparently this is also valid `tsql`, so I think dropping the `AS` during variable declaration is fine. 
```
CREATE PROCEDURE [dbo].[foo]
	@out_vchCode uddt_output_code
AS
BEGIN 
SELECT 'bar' 
END 
GO 
```

Additionally, with the removal of those keywords, the following results in no parse errors for me: 
```
SET QUOTED_IDENTIFIER ON
GO
SET ANSI_NULLS ON
GO

CREATE PROCEDURE [dbo].[usp_process_tran_log]
	  @out_vchCode uddt_output_code OUTPUT
	, @out_vchMsg uddt_output_msg OUTPUT
	, @in_debug INT = 1
AS
--*******************************************************************************************
SET NOCOUNT ON;
BEGIN 
SELECT '8'
END 
```
Note that this includes the `OUTPUT` hints.

**Comment by tunetheweb** [comment]
So question is, is the `AS` syntax valid and should we add it to SQLFluff? That wouldn’t be difficult but always an interesting call with these undocumented features even if they do work (they may stop working at any point as undocumented). WDYT?

I agree the `OUTPUT` does parse but it parses as a data type and that’s not right. So we should add this as it’s own syntax so that stops happening.

**Comment by edmoren** [comment]
@tunetheweb I'd leave it out since I cannot find any documentation and the main `sql-server` blog examples all exclude the `AS` in these instances. If you were to decide to include it, then I would classify the *presence* of that `AS` in variable declaration be an anti-pattern by default. I will be removing it across our code base when finding it as a result of this conversation. 

Possibly unrelated but I've come across another parse error with this `CREATE` statement. 
sample.sql 
```
SET QUOTED_IDENTIFIER ON
GO
SET ANSI_NULLS ON
GO

CREATE PROCEDURE [dbo].[usp_process_tran_log]
	  @out_vchCode uddt_output_code OUTPUT
	, @out_vchMsg uddt_output_msg OUTPUT
	, @in_debug INT = 1
AS
SET NOCOUNT ON; 	
BEGIN 
	BEGIN TRY
		SELECT '8';
	END TRY
	BEGIN CATCH
		SET @v_nSysErrorNum = ERROR_NUMBER();
		SET @v_vchCode = ERROR_LINE();
		SET @v_vchMsg = N'Missing control type.';
		SET @v_vchMsg = @v_vchMsg + N' SQL Error = ' + ERROR_MESSAGE();
		GOTO ERROR_HANDLER; -- this line causes parser error, removing or commenting it removes parser error.
	END CATCH
END 
```
parse output. 
```
==== parsing violations ====
L:  11 | P:  16 |  PRS | Line 11, Position 16: Found unparsable section: " \t\nBEGIN
                       | \n\tBEGIN TRY\n\t\tSELECT '8';\n\tEND ..."
```

Similar raw parse output to before: 
```
[L: 11, P: 15]      |                        [META] dedent:
[L: 11, P: 15]      |                        statement_terminator:                 ';'
[L: 11, P: 16]      |    unparsable:                                               !! Expected: 'Nothing...'
        Comments:
[L: 21, P: 23]      |            comment:                                          '-- this line 
causes parser error, removing or commenting it removes parser error.'
        Code:
[L: 11, P: 16]      |            whitespace:                                       ' \t'
[L: 11, P: 18]      |            newline:                                          '\n'
[L: 12, P:  1]      |            raw:                                              'BEGIN'     
```

**Comment by tunetheweb** [comment]
> After a little playing around with this, it looks like the GOTO keyword itself may not be supported.

Yup, that's the problem. Could you raise a separate issue for this?

**Comment by edmoren** [comment]
Sure, we can leave this open for the `OUTPUT` change.

**Comment by tunetheweb** [comment]
Fixed in #2949 if you wanna have a look at that?

**Comment by tunetheweb** [comment]
Also included the GOTO fix as dead simple. So no need to open a separate issue.

**Comment by edmoren** [comment]
Thanks @tunetheweb I will not submit a separate one. Taking a look at the PR.

**Comment by edmoren** [comment]
@tunetheweb left a small comment on your PR that's more of an inquisitory question on my part.
